### PR TITLE
[AMDGPU] Optimize the register uses if offset inlinable

### PR DESCRIPTION
**Pull request by PankajDwivedi-25**
Fold the frame index offset into v_mad if inlinable.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp [hunk 1]
```diff
@@ -2568,26 +2568,39 @@ bool SIRegisterInfo::eliminateFrameIndex(MachineBasicBlock::iterator MI,
...
+                // We start with a frame pointer with a wave space value, and
+                // an offset in lane-space. We are materializing a lane space
+                // value. We can either do a right shift of the frame pointer
+                // to get to lane space, or a left shift of the offset to get
+                // to wavespace. We can right shift after the computation to
+                // get back to the desired per-lane value. We are using the
+                // mad_u32_u24 primarily as an add with no carry out clobber.
+                bool IsInlinableLiteral = AMDGPU::isInlinableLiteral32(
+                    Offset, ST.hasInv2PiInlineImm());
+                if (!IsInlinableLiteral)
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp:2584]: Braces

File: llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp [hunk 2]
```diff
@@ -2568,26 +2568,39 @@ bool SIRegisterInfo::eliminateFrameIndex(MachineBasicBlock::iterator MI,
...
                 Add = BuildMI(*MBB, MI, DL, TII->get(AMDGPU::V_MAD_U32_U24_e64),
-                              TmpResultReg)
-                          .addReg(TmpResultReg, RegState::Kill)
-                          .addImm(ST.getWavefrontSize())
-                          .addReg(FrameReg)
-                          .addImm(0);
+                              TmpResultReg);
+
+                if (!IsInlinableLiteral) {
+                  Add.addReg(TmpResultReg, RegState::Kill)
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp:2593]: You can conditionally add this one operand and then share the rest of the .adds

**Review by cdevadas — APPROVED** [review]

File: llvm/test/CodeGen/AMDGPU/eliminate-frame-index-s-mov-b32.mir [hunk 3]
```diff
@@ -706,8 +706,7 @@ body:   |
     ; GFX8-NEXT: S_NOP 0, implicit-def $sgpr56_sgpr57_sgpr58_sgpr59_sgpr60_sgpr61_sgpr62_sgpr63
     ; GFX8-NEXT: V_CMP_EQ_U32_e32 0, killed $vgpr0, implicit-def $vcc, implicit $exec
     ; GFX8-NEXT: S_CMP_EQ_I32 $sgpr4, $sgpr5, implicit-def $scc
-    ; GFX8-NEXT: $vgpr0 = V_MOV_B32_e32 64, implicit $exec
-    ; GFX8-NEXT: $vgpr0 = V_MAD_U32_U24_e64 killed $vgpr0, 64, $sgpr32, 0, implicit $exec
+    ; GFX8-NEXT: $vgpr0 = V_MAD_U32_U24_e64 64, 64, $sgpr32, 0, implicit $exec
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/eliminate-frame-index-s-mov-b32.mir:709]: Might be a good idea to change the size of stack.0 to something different from the wavesize so it's clear we didn't get the right answer by accident

Comment (PankajDwivedi-25) [reply llvm/test/CodeGen/AMDGPU/eliminate-frame-index-s-mov-b32.mir:709]: that's really great point, adding another test.

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/eliminate-frame-index-s-mov-b32.mir:709]: Don't need to add one, just change the size of stack.0 to 60 or something

Comment (PankajDwivedi-25) [reply llvm/test/CodeGen/AMDGPU/eliminate-frame-index-s-mov-b32.mir:709]: just added

**Review by arsenm — APPROVED** [review]

**Pull request merged**